### DISCUSSION TITLE
How to calculate composite dice for multi-level segmentation instead of average dice

### DISCUSSION THREAD

**Discussion by Alex8211101** [Project-MONAI/tutorials#723] [Q&A]
When calculating the dice of organs of different sizes, I found that monai is to obtain the meandice of different organs, but the larger organs often occupy a larger pixel value. The calculation of meandice will affect the comprehensive value due to the lower dice of small organs. Is there any such method? Thank you in `advance!`

> `dice_metric = DiceMetric(include_background=True, reduction="mean")
dice_metric_batch = DiceMetric(include_background=True, _reduction="mean_batch")

These are my output dices.
_Epoch: 99, mean dice: 0.834375,femur dice: 0.949789,femoral cartilage dice: 0.736557,tibia dice:0.943555,tibia cartlage dice:0.707600_

**Comment by Nic-Ma** [comment]
Hi @Alex8211101 ,

Thanks for your interest here.
If I understand your question correctly, I think you can use `reduction="mean_channel"` to get mean dice of every image, then compute the weighted average by image sizes.
What do you think?

Thanks.

**Reply by Alex8211101** [reply]
Thank you very much for your quick reply, I've got the desired dice result, but now I have a new problem. Dice can reach 0.9+ during training, but the result of `evaluation on original image spacing` is not good, only about 0.2 or even 0.05, maybe you can provide some help? thank you very much.

**Reply by Nic-Ma** [reply]
Could you please help share your test program for further analysis?

Thanks.

**Reply by Alex8211101** [reply]
Of course！This is my inference code.
`def main():
    opt = Options().parse()

    test_images = sorted(glob(os.path.join(opt.images_folder, 'test', 'image*.nii')))
    test_segs = sorted(glob(os.path.join(opt.fumer_labels_folder, 'test', 'label*.nii')))
    test_dicts = [{'image': image_name, 'label': label_name}
                  for image_name, label_name in zip(test_images, test_segs)]
    test_org_transforms = Compose(
        [
            LoadImaged(keys="image"),
            EnsureChannelFirstd(keys='image'),
            Orientationd(keys=["image"], axcodes="RAS"),
            CropForegroundd(keys=["image"], source_key="image"),
            ScaleIntensityd(keys='image'),
            EnsureTyped(keys="image"),
        ]
    )
    test_org_ds = Dataset(
        data=test_dicts, transform=test_org_transforms)
    test_org_loader = DataLoader(test_org_ds, batch_size=1, num_workers=0)
    post_transforms = Compose([
        EnsureTyped(keys="pred"),
        Activationsd(keys="pred", softmax=True),
        Invertd(
            keys="pred",transform=test_org_transforms,orig_keys="image",meta_keys="pred_meta_dict",orig_meta_keys="image_meta_dict",
            meta_key_postfix="meta_dict",nearest_interp=False,to_tensor=True,
        ),
        AsDiscreted(keys="pred", argmax=True, to_onehot=2),
        SaveImaged(keys="pred",
                   meta_keys="pred_meta_dict",
                   output_dir="./out/femur/",
                   output_postfix="seg_femur", output_ext='.nii', resample=False),
        # AsDiscreted(keys="pred", argmax=True,to_onehot=5,),
        # SplitChanneld(keys="pred", channel_dim=0),
        # SaveImaged(keys="pred_0",
        #            meta_keys="pred_meta_dict",
        #            output_dir="./out/resUNet/",
        #            output_postfix="seg_0",output_ext='.nii' ,resample=False),
        # SaveImaged(keys="pred_1",
        #            meta_keys="pred_meta_dict",
        #            output_dir="./out/resUNet/",
        #            output_postfix="seg_1", output_ext='.nii', resample=False),
        # SaveImaged(keys="pred_2",
        #            meta_keys="pred_meta_dict",
        #            output_dir="./out/resUNet/",
        #            output_postfix="seg_2", output_ext='.nii', resample=False),
        # SaveImaged(keys="pred_3",
        #            meta_keys="pred_meta_dict",
        #            output_dir="./out/resUNet/",
        #            output_postfix="seg_3", output_ext='.nii', resample=False),
        # SaveImaged(keys="pred_4",
        #            meta_keys="pred_meta_dict",
        #            output_dir="./out/resUNet/",
        #            output_postfix="seg_4", output_ext='.nii', resample=False),
    ])

    from models.residual_unet3d import UNet
    net = UNet(in_channels=1, n_classes=5)
    # net = torch.nn.DataParallel(net)
    net.cuda()
    net.load_state_dict(torch.load(os.path.join('log/resUNet3d_femur', "best_metric_model.pth")))
    net.eval()
    with torch.no_grad():
        for test_data in test_org_loader:
            test_data['image'] = test_data["image"].cuda()
            roi_size = (320,320, 96)
            sw_batch_size = 1
            test_data["pred"] = sliding_window_inference(
                test_data['image'], roi_size, sw_batch_size, net)
            test_data = [post_transforms(i) for i in decollate_batch(test_data)]

if __name__ == "__main__":
    main()`

> Evaluation on original image spacing
`def main():
    opt = Options().parse()

    val_images = sorted(glob(os.path.join(opt.images_folder, 'val', 'image*.nii')))
    val_segs = sorted(glob(os.path.join(opt.femur_labels_folder, 'val', 'label*.nii')))
    val_dicts = [{'image': image_name, 'label': label_name}
                 for image_name, label_name in zip(val_images, val_segs)]

    val_org_transforms = Compose(
        [
            LoadImaged(keys=["image", "label"]),
            EnsureChannelFirstd(keys=["image", "label"]),
            Orientationd(keys=["image"], axcodes="RAS"),
            # Spacingd(keys=["image"], pixdim=(
            #     1.5, 1.5, 2.0), mode="bilinear"),
            # ScaleIntensityRanged(
            #     keys=["image"], a_min=-57, a_max=164,
            #     b_min=0.0, b_max=1.0, clip=True,
            # ),
            # CropForegroundd(keys=["image"], source_key="image"),
            EnsureTyped(keys=['image', 'label'])
        ]
    )

    val_org_ds = Dataset(
        data=val_dicts, transform=val_org_transforms)
    val_org_loader = DataLoader(val_org_ds, batch_size=1, num_workers=4)

    post_transforms = Compose([
        EnsureTyped(keys="pred"),
        Invertd(
            keys="pred",  # invert the `pred` data field, also support multiple fields
            transform=val_org_transforms,
            orig_keys="image",  
            meta_keys="pred_meta_dict", 
            orig_meta_keys="image_meta_dict", 
            meta_key_postfix="meta_dict",  
            nearest_interp=False, 
            to_tensor=True,
        ),
        AsDiscreted(keys="pred", argmax=True, to_onehot=2),
        AsDiscreted(keys="label", to_onehot=2),
    ])

    dice_metric = DiceMetric(include_background=False, reduction='mean')
    # net = UNet(dimensions=3, in_channels=1, out_channels=5, channels=(16, 32, 64, 128), strides=(2, 2, 2, 2),
    #            num_res_units=2, norm=Norm.BATCH
    #            )
    from models.residual_unet3d import UNet
    net = UNet(in_channels=1, n_classes=2)
    # net = torch.nn.DataParallel(net)
    net.cuda()

    net.load_state_dict(torch.load(
        os.path.join(
            'log/resUNet3d_femur', "best_metric_model.pth")))
    net.eval()

    with torch.no_grad():
        for val_data in val_org_loader:
            val_inputs = val_data["image"].cuda()
            roi_size = (256, 256, 48)
            sw_batch_size = 4
            val_data["pred"] = sliding_window_inference(
                val_inputs, roi_size, sw_batch_size, net)
            val_data = [post_transforms(i) for i in decollate_batch(val_data)]
            val_outputs, val_labels = from_engine(["pred", "label"])(val_data)
            # compute metric for current iteration
            dice_metric(y_pred=val_outputs, y=val_labels)

        # aggregate the final mean dice result
        metric_org = dice_metric.aggregate().item()
        # reset the status for next validation round
        dice_metric.reset()

    print("Metric on original image spacing: ", metric_org)

if __name__ == "__main__":
    main()`

**Reply by Nic-Ma** [reply]
Hi @Alex8211101 ,

I checked your program, seems you missed `Activationsd(keys="pred", softmax=True)` in your post transforms of evaluation?
Could you please help add it and try again?

Thanks.

**Reply by Alex8211101** [reply]
Thanks for the reminder,I made the modification as below.
`    post_transforms = Compose([
        EnsureTyped(keys="pred"),
        Invertd(
            keys="pred", 
            orig_keys="image",  
            meta_keys="pred_meta_dict", 
            orig_meta_keys="image_meta_dict",  
            meta_key_postfix="meta_dict",  
            nearest_interp=False, 
            to_tensor=True,
        ),
        Activationsd(keys='pred',softmax=True),
        AsDiscreted(keys="pred", argmax=True, to_onehot=2),
        # AsDiscreted(keys="label", to_onehot=2),
    ])`
But unfortunately I only got the result od dice=0.14.That's weird.I don't know if it's because of my wrong setting during traing that I get the segmentation results that looks good.Thank you very much for your help!
` dice_metric = DiceMetric(include_background=False, reduction='mean',)

    post_trans=Compose([EnsureType(),Activations(sigmoid=True),AsDiscrete(argmax=True,to_onehot=2)])
    post_label = Compose([EnsureType(), AsDiscrete(to_onehot=2)])

    loss_function = monai.losses.DiceCELoss(include_background=True, to_onehot_y=True, softmax=True)
    torch.backends.cudnn.benchmark = opt.benchmark

    val_interval = 1
    best_metric = -1
    best_metric_epoch = -1
    epoch_loss_values = []
    metric_values = []
    save_name_txt = os.path.join(opt.output_dir, "log_dir", 'Res_log_femur.txt')
    model_file = open(save_name_txt, "w")
    total_start=time.time()

    writer = SummaryWriter()
    for epoch in range(opt.epochs):
        epoch_start = time.time()
        print("-" * 10)
        print(f"epoch {epoch + 1}/{opt.epochs}")
        net.train()
        epoch_loss = 0
        step = 0
        for batch_data in train_loader:
            step += 1
            inputs, labels = batch_data["image"].cuda(), batch_data["label"].cuda()
            optim.zero_grad()
            outputs = net(inputs)
            # outputs=inferer(inputs.cuda(),net)
            loss = loss_function(outputs, labels)
            loss.backward()
            optim.step()
            epoch_loss += loss.item()
            epoch_len = len(check_train) // train_loader.batch_size
            print(f"{step}/{epoch_len}, train_loss: {loss.item():.4f}")
            writer.add_scalar("train_loss", loss.item(), epoch_len * epoch + step)
        epoch_loss /= step
        epoch_loss_values.append(epoch_loss)
        print(f"epoch {epoch + 1} average loss: {epoch_loss:.4f}")

        train_result = 'Epoch: %d, Train Loss: %f' % (epoch, epoch_loss)

        if (epoch + 1) % val_interval == 0:
            net.eval()
            with torch.no_grad():
                for val_data in val_loader:
                    val_inputs, val_labels = (
                        val_data["image"].cuda(),
                        val_data["label"].cuda(),
                    )
                    roi_size = (256, 256, 48)
                    sw_batch_size = 4
                    val_outputs = sliding_window_inference(
                        val_inputs, roi_size, sw_batch_size, net)
                    val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
                    val_labels = [post_label(i) for i in decollate_batch(val_labels)]
                    # compute metric for current iteration
                    dice_metric(y_pred=val_outputs, y=val_labels)


                    # aggregate the final mean dice result
                metric = dice_metric.aggregate().item()
                # reset the status for next validation round
                dice_metric.reset()
                # dice_metric_batch.reset()
                torch.save(net.state_dict(), os.path.join(
                    opt.root_dir, 'resUNet3d_femur',
                    "{}_Dice:{:.4f}.pth".format(epoch, metric)))
                metric_values.append(metric)
                if metric > best_metric:
                    best_metric = metric
                    best_metric_epoch = epoch + 1
                    torch.save(net.state_dict(), os.path.join(
                        'log/resUNet3d_femur', "best_metric_model.pth"))
                    print("saved new best metric model")
                print(
                    f"current epoch: {epoch + 1} current mean dice: {metric:.4f}" 
                    f"\nbest mean dice: {best_metric:.4f} "
                    f"at epoch: {best_metric_epoch}"
                )
                valid_result = 'Epoch: %d, mean dice: %4f' % (
                    epoch, metric)
        model_file.write(valid_result + '\n')
        model_file.write(train_result + '\n')
        print(f"time consuming of epoch{epoch + 1} is: {(time.time() - epoch_start):.4f} s")
    total_time = time.time() - total_start
    model_file.close()`

**Reply by Nic-Ma** [reply]
Please put the `Activationsd` before `Invertd`, just like your inference program.

Thanks.

**Reply by Alex8211101** [reply]
Thanks again for your reply, the monai team has been so helpful to me. I can already export the results of the segmentation, but there is one strange place. I input the `best metric model` for training the knee tibia, but the output turned out to be the segmentation result of the femur, and the accuracy of the segmentation is relatively high. I can confirm the data is accurate through the two-dimensional result of `check best model`.

**Reply by Nic-Ma** [reply]
I didn't get your problem "I input the best metric model for training the knee tibia, but the output turned out to be the segmentation result of the femur".
Maybe there is some confusion in your jupyter notebook? Could you please help totally restart your jupyter notebook and try again?

Thanks.
